### ISSUE TITLE
tables are a bit hard to follow in CtxInternal

### ISSUE THREAD

**Issue by Geal** [wasmerio/wasmer#1171]
I am experimenting with a flat `InternalCtx` structure (from runtime-core), where everything lives in the same buffer. The basic idea is to take an existing context, and generate a new one that will reside at a known address, and have every pointer refer to a later point in the buffer.
This has a few benefits:
- lighter cache usage: everything lives in contiguous data and will likely be prefetched
- easy to serialize: I can dump the buffer, and use it later for other executions (with caveats around imports though)

It's mostly working for now, except signature checking, which apparently uses tables (I have not covered element initializers yet). If I follow the generated code:
- load `InternalCtx`
- load the `tables` pointer
- load the pointed `LocalTable`. Apparently it is a contiguous list of `runtime_core::tables::Anyfunc`?
- `Anyfunc` contains a `AnyfuncInner` which is an enum that can contain a host function or a `DynFunc`. This is where I have a hard time following the generated code and lattening the structure, because `Anyfunc` is not `repr(C)` and contains an `Arc<FuncSig>`

If I understand correctly, it checks that the pointer to the signature of the function we're trying to call is the right one?
Is it necessary to have all those indirection levels? If I understand things correctly, it could be possible to heck that the function we're calling has the same signature index?

**Comment by syrusakbary** [comment]
Tables are completely refactored on the refactor. We should be able to solve this issue once it lands in master :)
